hasjob/search.py

```diff
@@ -1,83 +1,45 @@
 # -*- coding: utf-8 -*-
 
-import os.path
 from flask.ext.sqlalchemy import models_committed
 from flask.ext.rq import job
-import time
-from sqlalchemy.exc import ProgrammingError
-from whoosh import fields, index
-from whoosh.index import LockError
-from whoosh.qparser import QueryParser
-from whoosh.analysis import StemmingAnalyzer
-
 from hasjob import models, app
 
+from pyelasticsearch import ElasticSearch
+es = ElasticSearch('http://localhost:9200/')
+
+INDEXABLE = (models.JobPost,)
+ES_INDEX = 'hasjob'
 
-INDEXABLE = (models.JobType, models.JobCategory, models.JobPost)
-search_schema = fields.Schema(title=fields.TEXT(stored=True),
-                              content=fields.TEXT(analyzer=StemmingAnalyzer()),
-                              idref=fields.ID(stored=True, unique=True))
 
+def type_from_idref(idref):
+    return idref.split('/')[0]
 
-# For search results
-def ob_from_idref(idref):
-    ref, objid = idref.split('/')
-    if ref == 'post':
-        return models.JobPost.query.get(int(objid))
-    elif ref == 'type':
-        return models.JobType.query.get(int(objid))
-    elif ref == 'category':
-        return models.JobCategory.query.get(int(objid))
+
+def id_from_idref(idref):
+    return int(idref.split('/')[1])
 
 
 def do_search(query, expand=False):
     """
     Returns search results
     """
-    ix = index.open_dir(app.config['SEARCH_INDEX_PATH'])
-    hits = []
-
     if query:
-        parser = QueryParser("content", schema=ix.schema)
-        try:
-            qry = parser.parse(query)
-        except:
-            # query couldn't be parsed
-            qry = None
-        if qry is not None:
-            searcher = ix.searcher()
-            hits = searcher.search(qry, limit=1000)
-
-    if expand:
-        return (ob_from_idref(hit['idref']) for hit in hits)
-    else:
-        return hits
+        hits = es.search(query, index=ES_INDEX)
+        ids = map(lambda hit: id_from_idref(hit[u'_id']), hits['hits']['hits'])
+        job_post_ids = map(lambda post: post.id, models.JobPost.get_by_ids(ids))
```

Comment:
To make this efficient:

```python
job_post_ids = [r.id for r in db.session.query(models.JobPost.id).filter(models.JobPost.id.in_(ids))]
```

See? Doesn't involve a map+lambda (inefficient) or loading unwanted data from the database.